server/compose/types/record_detect_duplicates.go

```diff
@@ -160,49 +160,52 @@ func (rule DeDupRule) checkDuplication(ctx context.Context, ls localeService, re
-                // 2. multiValue is oneOf, then one or more value needs to be a match then return error/warning
-                // 3. multiValue is equal, then all value needs to be a match then return error/warning
-                if (!valErr.IsValid() && (!c.HasMultiValue() || c.IsAllEqual()) && valErr.Len() == rvv.Len()) || (c.IsOneOf() && valErr.Len() > 0) {
-                    if out == nil {
-                        out = &RecordValueErrorSet{}
+            out.Push(multiValErr)
+        } else {
+            _ = vv.Walk(func(v *RecordValue) error {
+                if v.RecordID != rec.ID {
+                    _ = rvv.Walk(func(rv *RecordValue) error {
+                        if len(rv.Value) > 0 && matchValue(c.Modifier, rv.Value, v.Value) {
+                            valErr.Push(RecordValueError{
+                                Kind:    rule.IssueKind(),
+                                Message: ls.T(ctx, "compose", rule.IssueMessage()),
+                                Meta: map[string]interface{}{
+                                    "field":         v.Name,
+                                    "value":         v.Value,
+                                    "dupValueField": rv.Name,
+                                    "recordID":      cast.ToString(v.RecordID),
+                                    "rule":          rule.String(),
+                                },
+                            })
+                        }
+                        return nil
+                    })
+
+                    // 1. multiValue is empty, then all value needs to be a match then return error/warning
+                    // 2. multiValue is oneOf, then one or more value needs to be a match then return error/warning
+                    // 3. multiValue is equal, then all value needs to be a match then return error/warning
+                    if (!valErr.IsValid() && (!c.HasMultiValue() || c.IsAllEqual()) && valErr.Len() == rvv.Len()) || (c.IsOneOf() && valErr.Len() > 0) {
+                        if out == nil {
+                            out = &RecordValueErrorSet{}
+                        }
+                        out.Push(valErr.Set...)
                     }
-                    out.Push(valErr.Set...)
                 }
-            }
-            return nil
-        })
+                return nil
+            })
+        }
     }
 
     return
 }
 
-func (rule DeDupRule) multiValueAllEqual(ctx context.Context, ls localeService, c *DeDupRuleConstraint, rvv RecordValueSet, existingVv RecordValueSet) (out *RecordValueErrorSet) {
-    var (
-        valErr = &RecordValueErrorSet{}
-    )
+func (rule DeDupRule) multiValueAllEqual(ctx context.Context, ls localeService, c *DeDupRuleConstraint, rvv RecordValueSet, existingVv RecordValueSet) (out RecordValueError) {
+    var valErr = RecordValueError{}
 
     if rvv.Len() == existingVv.Len() {
```

Comment:
you can make an early return with `if rvv.Len() != existingVv.Len()` so you wont have so much nesting